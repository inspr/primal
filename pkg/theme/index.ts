import { Unpack } from "@primal/state"
import Theme from "./src/theme"

// Apply the theme definitions to the global scope
// We need this definition to primal recognize the custom props
declare module "@primal/primitives" {
    /**
     * @type Inspr
     * @extends Spec
     * `@primal/theme` is currently applied as theme the app.
     *
     * The theme is designed and managed by inspr team and it's based on inspr design language
     * @link https://inspr.design/theme
     */

    type Theme = Unpack<typeof Theme>["__theme"]
    interface PrimitivesTheme extends Theme {}
}

export default Theme
