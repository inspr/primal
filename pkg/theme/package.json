{
    "name": "@primal/theme",
    "version": "0.0.0",
    "description": "",
    "main": "./index.ts",
    "types": "./index.ts",
    "dependencies": {
        "@types/next": "^9.0.0",
        "@types/react": "^17.0.39"
    }
}
