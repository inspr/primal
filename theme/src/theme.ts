import { createTheme } from "@primal/primitives"

const theme = createTheme({
    token: {
        color: {
            swatch: {
                brown: {
                    100: "#f8f3f1",
                    200: "#e4d0c9",
                    300: "#cfada0",
                    400: "#bb8977",
                    500: "#a36852",
                    600: "#7a4e3e",
                    700: "#513429",
                    800: "#36231b",
                    900: "#1b120e",
                },
                yellow: {
                    100: "#fff8eb",
                    200: "#ffe4ad",
                    300: "#ffcf70",
                    400: "#ffba33",
                    500: "#f5a300",
                    600: "#b87a00",
                    700: "#7a5200",
                    800: "#523600",
                    900: "#241800",
                },
                orange: {
                    100: "#fff1eb",
                    200: "#ffc9ad",
                    300: "#ffa070",
                    400: "#ff7733",
                    500: "#f55200",
                    600: "#b83d00",
                    700: "#7a2900",
                    800: "#521b00",
                    900: "#290d00",
                },
                purple: {
                    100: "#f5ebff",
                    200: "#d6adff",
                    300: "#b870ff",
                    400: "#9933ff",
                    500: "#7a00f5",
                    600: "#5c00b8",
                    700: "#3d007a",
                    800: "#290052",
                    900: "#140029",
                },
                red: {
                    100: "#ffebee",
                    200: "#ffadbb",
                    300: "#ff7088",
                    400: "#ff3355",
                    500: "#e50026",
                    600: "#b8001f",
                    700: "#7a0014",
                    800: "#52000e",
                    900: "#290007",
                },
                pink: {
                    100: "#feecf8",
                    200: "#fab2e2",
                    300: "#f679cd",
                    400: "#f231b2",
                    500: "#e60f9e",
                    600: "#ad0b77",
                    700: "#73074f",
                    800: "#4d0535",
                    900: "#26021a",
                },
                blue: {
                    100: "#ecf2fe",
                    200: "#b2cafb",
                    300: "#78a2f7",
                    400: "#3e7bf4",
                    500: "#0d56e7",
                    600: "#0a41ae",
                    700: "#072b74",
                    800: "#041d4d",
                    900: "#020f27",
                },
                green: {
                    100: "#effaf1",
                    200: "#c1ecc6",
                    300: "#92dd9b",
                    400: "#63cf70",
                    500: "#3abb49",
                    600: "#2b8c36",
                    700: "#1d5e24",
                    800: "#133e18",
                    900: "#0a1f0c",
                },
                gray: {
                    50: "#fefefe",
                    100: "#f5f5f5",
                    200: "#d6d6d6",
                    300: "#b8b8b8",
                    400: "#999999",
                    500: "#7a7a7a",
                    600: "#5c5c5c",
                    700: "#3d3d3d",
                    800: "#292929",
                    900: "#111111",
                },
            },
            kind: {
                lumen: "gray.50",
                umbra: "gray.900",
                creativity: "yellow.400",
                danger: "red.500",
                privacy: "purple.400",
                trust: "blue.500",
            },
            gradient: {
                cta: "linear-gradient(90deg, #fc5d32 0%, #ee4c54 25%, #d45e95 50%, #9c6ca6 75%, #6583c1 100%)",
            },
        },
        font: {
            family: {
                inter: '"Inter", sans-serif',
                opensauce: '"Open Sauce One", sans-serif',
            },
            weight: {
                regular: "400",
                medium: "500",
                semibold: "600",
                bold: "700",
            },
            letterSpace: {
                sm: "-0.02rem",
                md: "0.01rem",
                lg: "0.04rem",
            },
            lineHeight: {
                xxxs: "12px",
                xxs: "16px",
                xs: "20px",
                sm: "24px",
                md: "28px",
                lg: "36px",
                xl: "44px",
                xxl: "64px",
                xxxl: "80px",
            },
            size: {
                xxxs: "10px",
                xxs: "12px",
                xs: "14px",
                sm: "16px",
                md: "18px",
                lg: "28px",
                xl: "36px",
                xxl: "48px",
                xxxl: "64px",
            },
        },
        shadow: {
            xs: "0px 2px 4px rgba(17, 17, 17, 0.04), 0px 1px 2px rgba(17, 17, 17, 0.04)",
            sm: "0px 2px 4px rgba(17, 17, 17, 0.05), 0px 2px 4px rgba(17, 17, 17, 0.05)",
            md: "0px 4px 6px rgba(17, 17, 17, 0.04), 0px 6px 12px 2px rgba(17, 17, 17, 0.05)",
            lg: "0px 16px 32px 2px rgba(17, 17, 17, 0.08), 0px 32px 48px rgba(17, 17, 17, 0.08)",
            xl: "0px 48px 64px 2px rgba(17, 17, 17, 0.1), 0px 64px 80px rgba(17, 17, 17, 0.1)",
        },
        size: {
            xxxs: "2px",
            xxs: "4px",
            xs: "8px",
            sm: "16px",
            md: "24px",
            lg: "32px",
            xl: "48px",
            xxl: "64px",
            xxxl: "96px",
        },
        breakpoint: {
            xs: { min: "0px", max: "600px" },
            sm: { min: "600px", max: "900px" },
            md: { min: "900px", max: "1280px" },
            lg: { min: "1280px", max: "1920px" },
            xl: { min: "1920px", max: "10000px" },
        },
    },
    variants: {
        text: {
            default: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "400",
                fontSize: "16px",
                lineHeight: "24px",
                fontFamily: '"Inter", sans-serif',
            },
            caption: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "400",
                fontSize: "12px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            display: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "700",
                fontSize: "48px",
                lineHeight: "44px",
                fontFamily: '"Open Sauce One", sans-serif',
            },
            overline: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "600",
                fontSize: "12px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            subtitle: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "500",
                fontSize: "16px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            title: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "700",
                fontSize: "28px",
                lineHeight: "36px",
                fontFamily: '"Open Sauce One", sans-serif',
            },
        },
        button: {
            primary: {
                textColor: "gray.50",
                backgroundColor: "gray.900",
                borderColor: "gray.900",
                iconColor: "invert(1)",
                hover: {
                    textColor: "gray.900",
                    backgroundColor: "gray.50",
                    iconColor: "invert(0)",
                },
            },
            secondary: {
                textColor: "gray.900",
                backgroundColor: "gray.50",
                borderColor: "gray.900",
                hover: {
                    textColor: "gray.50",
                    backgroundColor: "gray.900",
                    iconColor: "invert(1)",
                },
            },
            success: {
                textColor: "gray.50",
                backgroundColor: "blue.500",
                borderColor: "blue.500",
                iconColor: "invert(1)",
                hover: {
                    textColor: "blue.500",
                    backgroundColor: "gray.50",
                    iconColor:
                        "invert(17%) sepia(100%) saturate(4267%) hue-rotate(221deg) brightness(97%) contrast(90%)",
                },
            },
            error: {
                textColor: "gray.50",
                backgroundColor: "red.500",
                borderColor: "red.500",
                iconColor: "invert(1)",
                hover: {
                    textColor: "red.500",
                    backgroundColor: "gray.50",
                    iconColor:
                        "invert(24%) sepia(50%) saturate(6429%) hue-rotate(338deg) brightness(81%) contrast(128%)",
                },
            },
            disabled: {
                textColor: "gray.600",
                backgroundColor: "gray.400",
                borderColor: "gray.400",
                iconColor:
                    "invert(38%) sepia(1%) saturate(0%) hue-rotate(42deg) brightness(90%) contrast(88%)",
            },
            cta: {
                textColor: "lumen",
                backgroundColor: "cta",
                borderColor: "transparent",
                shadow: "md",
                iconColor: "invert(1)",
                hover: {
                    shadow: "lg",
                    opacity: {
                        value: 0.35,
                        color: "umbra",
                    },
                },
            },
        },
        switch: {
            primary: {
                onColor: "umbra",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            secondary: {
                onColor: "gray.500",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            success: {
                onColor: "blue.500",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            error: {
                onColor: "danger",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
        },
        checkbox: {
            primary: {
                backgroundColor: "gray.900",
                borderColor: "gray.900",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "umbra",
                },
            },
            secondary: {
                backgroundColor: "lumen",
                borderColor: "umbra",
                checked: {
                    backgroundColor: "umbra",
                    checkColor: "lumen",
                },
            },
            success: {
                backgroundColor: "blue.500",
                borderColor: "blue.500",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "blue.500",
                },
            },
            error: {
                backgroundColor: "red.500",
                borderColor: "red.500",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "red.500",
                },
            },
        },
        radio: {
            primary: {
                textColor: "umbra",
                backgroundColor: "gray.900",
                borderColor: "gray.900",
            },
            success: {
                textColor: "umbra",
                backgroundColor: "blue.500",
                borderColor: "blue.500",
            },
            error: {
                textColor: "umbra",
                backgroundColor: "red.500",
                borderColor: "red.500",
            },
            disabled: {
                textColor: "gray.600",
                backgroundColor: "gray.600",
                borderColor: "gray.600",
            },
        },
    },
})

export default theme
