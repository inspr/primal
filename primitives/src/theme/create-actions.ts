import { AnySize, Swatch, VariantOptions, Weight } from "../types"
import { Spec, Token } from "./index"

//TODO check types that are "string | number" e.g "presetValueAux"

export const createActions = (theme: Spec) => ({
    color(colorValue: string | `${string}.${Weight}`) {
        if (!colorValue || colorValue === "transparent") return colorValue

        const { swatch, kind, gradient } = theme.token.color

        if (Object.keys(gradient).includes(colorValue)) {
            return gradient[colorValue]
        }

        let colorValueAux = colorValue

        if (Object.keys(kind).includes(colorValue)) {
            colorValueAux = kind[colorValue]
        }
        const [color, weight] = colorValueAux.split(".") as [
            keyof typeof swatch,
            Weight
        ]
        const selectedColor: { [weight in Weight]?: Swatch } = swatch[color]
        // TODO: fix typing so this validation isn't necessary (ex: brown.50 seems to be valid)
        const validColor = selectedColor && selectedColor[weight]
        return validColor ? selectedColor[weight]! : "transparent"
    },
    fontWeight(weightValue: string) {
        const { weight } = theme.token.font
        let weightOptions = weightValue as keyof typeof weight
        return weight[weightOptions]
    },
    fontFamily(familyValue: string) {
        const { family } = theme.token.font
        let familyOptions = familyValue as keyof typeof family
        return family[familyOptions]
    },
    fontSize(sizeValue: AnySize) {
        const { size } = theme.token.font
        let sizeOptions = sizeValue as keyof typeof size
        return size[sizeOptions]
    },
    letterSpace(spaceValue: AnySize) {
        const { letterSpace } = theme.token.font
        // const letterSpace = theme.token.font["letterSpace"]
        let spaceOptions = spaceValue as keyof typeof letterSpace
        return letterSpace[spaceOptions]
    },
    lineHeight(heightValue: AnySize) {
        const { lineHeight } = theme.token.font
        let heightOptions = heightValue as keyof typeof lineHeight
        return lineHeight[heightOptions]
    },
    size(sizeValue: AnySize) {
        const { size } = theme.token
        const sizeOptions = sizeValue as keyof typeof size
        return size[sizeOptions]
    },
    shadow(shadowValue: AnySize) {
        const { shadow } = theme.token
        const shadowOptions = shadowValue as keyof typeof shadow
        return shadow[shadowOptions]
    },
    breakpoints() {
        return theme.token.breakpoint as Token["breakpoint"]
    },
    variants(component: keyof VariantOptions, variant: string) {
        const { variants } = theme
        const selected = variants[component]
        const type = variant as keyof typeof selected

        return variants[component][type]
    },
})

export type ActionType = ReturnType<typeof createActions>
