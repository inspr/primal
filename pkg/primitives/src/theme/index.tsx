import type { PrimitivesTheme } from "@primal/primitives"
import { AtomicState, createState, Unpack, useInspr } from "@primal/state"
import {
    AnyShadow,
    AnyTypedSize,
    Color,
    FontFamily,
    PX,
    Size,
    Unit,
    Weight,
} from "../types"
import { VariantOptions } from "../types/variants"
import { createActions } from "./create-actions"

interface DefaultTheme {
    token: {
        shadow: { [size in Size]: string }
        color: {
            swatch: {}
            kind: {}
            gradient: {}
        }
        font: {
            family: {}
            lineHeight: {}
            weight: {}
            letterSpace: {}
            size: {}
        }
        size: {}
        breakpoint: {}
    }
    variants: {}
}

type Theme = DefaultTheme & PrimitivesTheme
export type Token = Theme["token"]
export type Variants = Theme["variants"]

export interface Spec {
    token: {
        color: Color
        font: {
            family: {
                [key: string]: FontFamily
            }
            weight: {
                [key: string]: Weight
            }
            letterSpace: {
                [size in Size<0>]: Unit
            }
            lineHeight: AnyTypedSize<Unit>
            size: AnyTypedSize<PX>
        }
        shadow: AnyTypedSize<AnyShadow>
        size: AnyTypedSize<Unit>
        breakpoint: AnyTypedSize<{ min: PX; max: PX }>
    }
    variants: VariantOptions
}

let primitivesThemeContext: AtomicState<any>

export const createTheme = <T extends Spec>(spec: T) => {
    const theme = createState({
        __theme: spec,
        actions: createActions(spec),
    })

    primitivesThemeContext = theme
    return theme
}

export const useTheme = (): Unpack<ReturnType<typeof createTheme>> => {
    if (!primitivesThemeContext) {
        if (process.env.NODE_ENV === "development") {
            throw new Error("A theme must be registered before use")
        }
    }
    return useInspr(primitivesThemeContext)
}
