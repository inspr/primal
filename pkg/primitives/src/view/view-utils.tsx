import { css, SerializedStyles } from "@emotion/react"
import styled from "@emotion/styled"
import { Token } from "../theme"
import { ActionType } from "../theme/create-actions"
import {
    AllSizes,
    DimensionValue,
    LayoutProps,
    StyleProps,
    ViewBreakPoints,
} from "../types"

export const validateSmallestBPUsage = (
    usedBreakpoints: ViewBreakPoints,
    allBreakPoints: Token["breakpoint"]
) => {
    const usedBPKeys = Object.keys(usedBreakpoints)
    const smallestBP = findSmallestBreakPoint(
        Object.keys(allBreakPoints)
    ) as keyof Token["breakpoint"]

    if (usedBPKeys.includes(smallestBP))
        console.warn(
            "You are using the smallest breakpoint defined on theme.ts as a prop, " +
                "this breakpoint will be ignored by the View component. " +
                "Use the 'layout' and 'style' props instead."
        )
}

// replaceCSSProp is a helper function that gets the respective size values for props
const replaceCSSProp = (
    value: DimensionValue,
    cssProp: string,
    actions: ActionType
) => {
    if (!value) return ""

    const staticSizes = [
        "sm",
        "md",
        "lg",
        "xs",
        "xl",
        "xxs",
        "xxl",
        "xxxs",
        "xxxl",
    ]
    return `${cssProp}: ${
        staticSizes.includes(value) ? actions.size(value as AllSizes) : value
    };`
}

// findSmallestBreakPoint returns the smallest breakpoint of the current Theme
const findSmallestBreakPoint = (breakpoints: string[]) => {
    if (breakpoints.includes("xxxs")) return "xxxs"
    if (breakpoints.includes("xxs")) return "xxs"
    if (breakpoints.includes("xs")) return "xs"
    return "sm"
}

// generateCSS receives layout and style values and a breakpoint, and generates the
// media query for it
const generateMediaQueryCSS = (
    actions: ActionType,
    usedBreakpoints: ViewBreakPoints
) => {
    const allBreakpoints = actions.breakpoints()
    type BreakpointKeys = keyof typeof usedBreakpoints
    const bpKeys = Object.keys(usedBreakpoints) as BreakpointKeys[]

    let mediaStyles: SerializedStyles[] = []

    bpKeys.forEach((bp) => {
        const mediaCSS = css`
            @media only screen and (min-width: ${allBreakpoints[bp].min}) {
                ${mountCSSString(
                    actions,
                    usedBreakpoints[bp]?.layout,
                    usedBreakpoints[bp]?.styles
                )}
            }
        `
        mediaStyles.push(mediaCSS)
    })

    return mediaStyles
}

const mountCSSString = (
    actions: ActionType,
    layout?: LayoutProps,
    styles?: StyleProps
) => {
    if (!layout && !styles) return ""

    return `${
        layout?.display && layout?.display !== "none"
            ? `display: ${layout?.display};`
            : ""
    }
        ${replaceCSSProp(
            layout?.padding?.top as DimensionValue,
            "padding-top",
            actions
        )}
        ${replaceCSSProp(
            layout?.padding?.right as DimensionValue,
            "padding-right",
            actions
        )}
        ${replaceCSSProp(
            layout?.padding?.bottom as DimensionValue,
            "padding-bottom",
            actions
        )}
        ${replaceCSSProp(
            layout?.padding?.left as DimensionValue,
            "padding-left",
            actions
        )}
        ${replaceCSSProp(
            layout?.margin?.top as DimensionValue,
            "margin-top",
            actions
        )}
        ${replaceCSSProp(
            layout?.margin?.right as DimensionValue,
            "margin-right",
            actions
        )}
        ${replaceCSSProp(
            layout?.margin?.bottom as DimensionValue,
            "margin-bottom",
            actions
        )}
        ${replaceCSSProp(
            layout?.margin?.left as DimensionValue,
            "margin-left",
            actions
        )}
        ${replaceCSSProp(layout?.width as DimensionValue, "width", actions)}
        ${replaceCSSProp(
            layout?.maxWidth as DimensionValue,
            "max-width",
            actions
        )}
        ${replaceCSSProp(
            layout?.minWidth as DimensionValue,
            "min-width",
            actions
        )}
        ${replaceCSSProp(layout?.height as DimensionValue, "height", actions)}
        ${replaceCSSProp(layout?.top as DimensionValue, "top", actions)}
        ${replaceCSSProp(layout?.bottom as DimensionValue, "bottom", actions)}
        ${replaceCSSProp(layout?.left as DimensionValue, "left", actions)}
        ${replaceCSSProp(layout?.right as DimensionValue, "right", actions)}
        ${replaceCSSProp(
            layout?.maxHeight as DimensionValue,
            "max-height",
            actions
        )}
        ${replaceCSSProp(
            layout?.minHeight as DimensionValue,
            "min-height",
            actions
        )}
        ${layout?.overflow ? `overflow: ${layout?.overflow};` : ""}
        ${layout?.zIndex ? `z-index: ${layout?.zIndex};` : ""}
        ${layout?.position ? `position: ${layout?.position};` : ""}
        ${layout?.gap ? `gap : ${layout?.gap};` : ""}
        ${
            styles?.shadow
                ? `box-shadow : ${actions.shadow(styles?.shadow)};`
                : ""
        }
        ${
            styles?.backgroundColor
                ? `background-color: ${actions.color(styles?.backgroundColor)};`
                : ""
        }
        ${
            styles?.backgroundImage
                ? `background-image : ${styles?.backgroundImage};`
                : ""
        }
        ${
            styles?.backgroundPosition
                ? `background-position: ${styles?.backgroundPosition};`
                : ""
        }
        ${
            styles?.backgroundRepeat
                ? `background-repeat: ${styles?.backgroundRepeat};`
                : ""
        }
        ${
            styles?.backgroundSize
                ? `background-size: ${styles?.backgroundSize};`
                : ""
        }
        ${
            styles?.borderColor
                ? `border-color: ${actions.color(styles?.borderColor)};`
                : ""
        }
        ${styles?.borderRadius ? `border-radius: ${styles?.borderRadius};` : ""}
        ${styles?.borderWidth ? `border-width: ${styles?.borderWidth};` : ""}
        ${styles?.borderStyle ? `border-style: ${styles?.borderStyle};` : ""}
        ${styles?.opacity ? `opacity: ${styles?.opacity};` : ""}
        ${styles?.cursor ? `cursor: ${styles?.cursor};` : ""}
        ${styles?.transition ? `transition: ${styles?.transition};` : ""}`
}

export const ViewStyle = styled.div<{
    actions: ActionType
    usedBreakpoints: ViewBreakPoints
    layout?: LayoutProps
    styles?: StyleProps
    injectedCSS?: SerializedStyles
}>`
    ${(props) => mountCSSString(props.actions, props.layout, props.styles)}
    ${(props) => generateMediaQueryCSS(props.actions, props.usedBreakpoints)}
    ${(props) => (props.injectedCSS ? props.injectedCSS : "")}
`
