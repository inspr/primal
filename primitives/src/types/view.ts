import { Token } from "../theme"
import { onEvents } from "./event"
import { LayoutProps } from "./layout"
import { StyleProps } from "./style"

export type ViewBreakPoints = {
    [key in keyof Token["breakpoint"]]?: {
        layout?: LayoutProps
        styles?: StyleProps
    }
}

export interface ViewProps extends ViewBreakPoints {
    layout?: LayoutProps
    styles?: StyleProps
    onEvents?: onEvents
    // Each time you use this prop, a panda dies
    __unsafeClassName?: string
}
