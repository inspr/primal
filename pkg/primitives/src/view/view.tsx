import React, { FunctionComponent } from "react"
import { useTheme } from "../theme"
import { ViewBreakPoints, ViewProps } from "../types"
import { BaseProps } from "../utils"
import { validateSmallestBPUsage, ViewStyle } from "./view-utils"

export type ViewKinds =
    | "div"
    | "main"
    | "article"
    | "footer"
    | "nav"
    | "a"
    | "aside"
    | "body"
    | "canvas"
    | "button"
    | "section"
    | "ul"
    | "ol"
    | "span"

const View: FunctionComponent<BaseProps<ViewKinds> & ViewProps> = ({
    is = "div",
    children,
    layout,
    styles,
    onEvents: { onClick, onChange } = {},
    injectedCSS,
    __unsafeClassName,
    ...props
}) => {
    const { actions } = useTheme()
    const usedBreakpoints = { ...props } as ViewBreakPoints

    if (process.env.NODE_ENV === "development")
        validateSmallestBPUsage(usedBreakpoints, actions.breakpoints())

    return (
        <ViewStyle
            as={is}
            actions={actions}
            usedBreakpoints={usedBreakpoints}
            layout={layout}
            styles={styles}
            injectedCSS={injectedCSS}
            className={__unsafeClassName}>
            {children}
        </ViewStyle>
    )
}

View.displayName = "View"

export { View }
