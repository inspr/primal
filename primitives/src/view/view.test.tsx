import "@primal/theme"
import { create } from "react-test-renderer"
import { View } from "./view"

describe("View", () => {
    it("should match snapshot", () => {
        const view = create(<View />)
        expect(view.toJSON()).toMatchSnapshot()
    })

    it("renders a custom View", () => {
        const view = create(
            <View
                is="section"
                layout={{
                    top: "50px",
                    right: "50px",
                    bottom: "50px",
                    left: "50px",
                    overflow: "hidden",
                    display: "flex",
                    padding: {
                        top: "xxxs",
                        right: "xxs",
                        bottom: "xl",
                        left: "xxl",
                    },
                    margin: {
                        top: "xxxs",
                        right: "xxs",
                        bottom: "xl",
                        left: "xxl",
                    },
                    zIndex: 1,
                    position: "absolute",
                    width: "sm",
                    maxWidth: "xxxs",
                    minWidth: "xs",
                    height: "md",
                    maxHeight: "xxxs",
                    minHeight: "xl",
                    gap: "5px",
                }}
                styles={{
                    shadow: "xl",
                    backgroundColor: "danger",
                    backgroundImage:
                        "linear-gradient(180deg, #EDEEF1 18.05%, rgba(255, 255, 255, 0) 100%)",
                    backgroundPosition: "top",
                    backgroundRepeat: "repeat-x",
                    backgroundSize: "contain",
                    borderColor: "gray.200",
                    borderRadius: "5px",
                    borderWidth: "5px",
                    borderStyle: "dashed",
                    opacity: 1,
                    cursor: "pointer",
                    transition: "width 2s",
                }}
                md={{
                    styles: {
                        borderStyle: "solid",
                    },
                    layout: {
                        width: "fit-content",
                    },
                }}
                __unsafeClassName="someClassName">
                Some View
            </View>
        )
        const props = view.root.findByProps({ is: "section" }).props

        // Verify objects
        expect(props).toHaveProperty("children", "Some View")
        expect(props).toHaveProperty("__unsafeClassName", "someClassName")
        expect(props).toHaveProperty("layout")
        expect(props).toHaveProperty("styles")
        expect(props).toHaveProperty("md")

        // Verify layout props
        expect(props.layout).toHaveProperty("display", "flex")
        expect(props.layout).toHaveProperty("padding", {
            top: "xxxs",
            right: "xxs",
            bottom: "xl",
            left: "xxl",
        })
        expect(props.layout).toHaveProperty("margin", {
            top: "xxxs",
            right: "xxs",
            bottom: "xl",
            left: "xxl",
        })
        expect(props.layout).toHaveProperty("zIndex", 1)
        expect(props.layout).toHaveProperty("position", "absolute")
        expect(props.layout).toHaveProperty("width", "sm")
        expect(props.layout).toHaveProperty("maxWidth", "xxxs")
        expect(props.layout).toHaveProperty("minWidth", "xs")
        expect(props.layout).toHaveProperty("height", "md")
        expect(props.layout).toHaveProperty("maxHeight", "xxxs")
        expect(props.layout).toHaveProperty("minHeight", "xl")
        expect(props.layout).toHaveProperty("gap", "5px")
        expect(props.layout).toHaveProperty("top", "50px")
        expect(props.layout).toHaveProperty("right", "50px")
        expect(props.layout).toHaveProperty("bottom", "50px")
        expect(props.layout).toHaveProperty("left", "50px")
        expect(props.layout).toHaveProperty("overflow", "hidden")
        // Verify style props
        expect(props.styles).toHaveProperty("shadow", "xl")
        expect(props.styles).toHaveProperty("backgroundColor", "danger")
        expect(props.styles).toHaveProperty(
            "backgroundImage",
            "linear-gradient(180deg, #EDEEF1 18.05%, rgba(255, 255, 255, 0) 100%)"
        )
        expect(props.styles).toHaveProperty("backgroundPosition", "top")
        expect(props.styles).toHaveProperty("backgroundRepeat", "repeat-x")
        expect(props.styles).toHaveProperty("backgroundSize", "contain")
        expect(props.styles).toHaveProperty("borderColor", "gray.200")
        expect(props.styles).toHaveProperty("borderRadius", "5px")
        expect(props.styles).toHaveProperty("borderWidth", "5px")
        expect(props.styles).toHaveProperty("borderStyle", "dashed")
        expect(props.styles).toHaveProperty("opacity", 1)
        expect(props.styles).toHaveProperty("cursor", "pointer")
        expect(props.styles).toHaveProperty("transition", "width 2s")

        // Verify media query (md) props
        expect(props.md).toHaveProperty("layout")
        expect(props.md).toHaveProperty("styles")
        expect(props.md.layout).toHaveProperty("width", "fit-content")
        expect(props.md.styles).toHaveProperty("borderStyle", "solid")
    })
})
