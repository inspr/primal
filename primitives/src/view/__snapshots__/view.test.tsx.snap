// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View should match snapshot 1`] = `
<div
  className="css-rvandl"
/>
`;
